### PR TITLE
Feat/create column meta data

### PR DESCRIPTION
**Pull request by lucafurrer**
Create a dimensionMetadata entry whenever a column mapping is created

Part of #187

### CODE REVIEW

**Comment by martinmaillard** [comment]
One remark: we only need to create `DimensionMetadata` for column mappings on the `ObservationTable`(s)

File: apis/core/lib/domain/column-mapping/create.ts [hunk 1]
```diff
@@ -39,6 +42,16 @@ export async function createColumnMapping({
     defaultValue: resource.out(cc.defaultValue).term,
   })
 
+  const dimensionMetaDataCollectionPointer = await getDimensionMetaDataCollection(table.csvMapping.id)
+  const dimensionMetaDataCollection = await store.getResource<DimensionMetadataCollection>(dimensionMetaDataCollectionPointer)
+  if (!dimensionMetaDataCollection) {
+    throw new Error(`DimensionMetadataCollection ${dimensionMetaDataCollectionPointer} not found`)
```
Comment (tpluscode) [inline apis/core/lib/domain/column-mapping/create.ts:48]: You can simpler
```suggestion
    throw new NotFoundError(`{dimensionMetaDataCollectionPointer`)
```

File: apis/core/lib/domain/identifiers.ts [hunk 2]
```diff
@@ -55,3 +55,7 @@ export function user(user: string): NamedNode {
 export function job(jobCollection: GraphPointer<NamedNode>): NamedNode {
   return $rdf.namedNode(`${jobCollection.value}/${nanoid()}`)
 }
+
+export function dimensionMetadata(dimensionMetadataCollection: DimensionMetadataCollection, mapping: ColumnMapping): NamedNode {
+  return $rdf.namedNode(`${dimensionMetadataCollection.pointer}?dimension=${mapping.pointer}`)
```
Comment (tpluscode) [inline apis/core/lib/domain/identifiers.ts:60]: We discussed with @martinmaillard and query params are not such a great choice after all.

Check the updated seed graph in [#200](https://github.com/zazuko/cube-creator/pull/200/files#diff-ee0d6661710ff05549447fce5560bf3fde360e65f8ea766d8a1ab77c1e40ebe9R50-R54)

I propose you simply concatenate the collection ID with the `targetProperty`. Keep the string value unchanged and maybe use `import { shrink } from '@zazuko/rdf-vocabularies'` to shorten URI target properties

File: apis/core/lib/domain/dimension-metadata/DimensionMetadata.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  return class Impl extends Resource implements ApiDimensionMetadataCollection {
+    addDimensionMetadata(params: CreateColumnMetadata): DimensionMetadata.DimensionMetadata {
+      const dimensionMetadata = DimensionMetadata.create(this.pointer.node(id.dimensionMetadata(this, params.columnMapping)), {
+        about: createIdentifier(params.columnMapping.targetProperty),
+      })
+
+      this.pointer.addOut(schema.hasPart, dimensionMetadata.id)
+
+      return dimensionMetadata
+    }
```
Comment (tpluscode) [inline apis/core/lib/domain/dimension-metadata/DimensionMetadata.ts:37]: Please throw if another metadata resource already exists for the same`schema:about`

File: apis/core/lib/domain/queries/dimension-metadata.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+        GRAPH ?dataset
+        {
+            ?dataset ${cc.dimensionMetadata} ?dimensionMetadata .
+        }
+        `
+    .LIMIT(1)
+    .execute(client.query)
+  if (results.length === 0) {
+    throw new Error(`No DimensionMetadata for table ${csvMapping} found`)
+  }
```
Comment (tpluscode) [inline apis/core/lib/domain/queries/dimension-metadata.ts:25]: Why `LIMIT`?
I would actually throw for anything which is `length !== 1`

File: apis/core/lib/domain/table/create.ts [hunk 5]
```diff
@@ -52,6 +55,12 @@ export async function createTable({
   })
 
   // Create default column mappings for provided columns
+  const dimensionMetaDataCollectionPointer = await getDimensionMetaDataCollection(csvMapping.id)
+  const dimensionMetaDataCollection = await store.getResource<DimensionMetadataCollection>(dimensionMetaDataCollectionPointer)
+  if (!dimensionMetaDataCollection) {
+    throw new Error(`DimensionMetadataCollection ${dimensionMetaDataCollectionPointer} not found`)
```
Comment (tpluscode) [inline apis/core/lib/domain/table/create.ts:61]: Ditto

```suggestion
    throw new NotFoundError(dimensionMetaDataCollectionPointer)
```

File: packages/model/DimensionMetadata.ts [hunk 6]
```diff
@@ -4,14 +4,23 @@ import { schema } from '@tpluscode/rdf-ns-builders'
...
 
+export function DimensionMetadataMixin<Base extends Constructor>(Resource: Base) {
+  class Impl extends Resource implements Partial<DimensionMetadata> {
+    @property.resource({ path: schema.about })
+    about!: NamedNode
+  }
+
+  return Impl
+}
+
```
Comment (tpluscode) [inline packages/model/DimensionMetadata.ts:23]: re. the CI error, look you will need to annotate the intended target for this mixin. This one does not have a class associated nor any other way to use it implicitly. Maybe try this

```suggestion
export function DimensionMetadataMixin<Base extends Constructor>(Resource: Base) {
  class Impl extends Resource implements Partial<DimensionMetadata> {
    @property.resource({ path: schema.about })
    about!: NamedNode
  }

  return Impl
}

DimensionMetadataMixin.shouldApply = false
```

File: apis/core/test/domain/column-mapping/create.test.ts [hunk 7]
```diff
@@ -104,4 +116,23 @@ describe('domain/column-mapping/create', () => {
...
+
+    // when
+    await createColumnMapping({ resource, store, tableId: table.term, dimensionMetadataQueries })
+
+    // then
+    expect(dimensionMetadata).to.matchShape({
+      property: [{
+        path: schema.hasPart,
+        minCount: 1,
+      }],
```
Comment (tpluscode) [inline apis/core/test/domain/column-mapping/create.test.ts:135]: You can also use `sh:node` to check deeper into the graph

```suggestion
      property: [{
        path: schema.hasPart,
        minCount: 1,
        node: {
          property: {
            path: schema.about,
            minCount: 1,
            maxCount: 1,
          }
        }
      }],
```

File: apis/core/test/domain/table/create.test.ts [hunk 8]
```diff
@@ -153,9 +164,33 @@ describe('domain/table/create', () => {
...
+
+    // when
+    await createTable({ resource, store, tableCollection, dimensionMetadataQueries })
+
+    // then
+    expect(dimensionMetadata).to.matchShape({
+      property: [{
+        path: schema.hasPart,
+        minCount: 2,
+      }],
```
Comment (tpluscode) [inline apis/core/test/domain/table/create.test.ts:193]: Ditto. And btw, `property` does not have to an array

File: apis/core/lib/domain/dataset/update.ts [hunk 9]
```diff
@@ -41,6 +43,7 @@ export async function update({
     .addOut(rdf.type, dcat.Dataset)
 
   hasPart.forEach(child => datasetResource.addOut(schema.hasPart, child))
+  dimensionMetadata.forEach(child => datasetResource.addOut(cc.dimensionMetadata, child))
```
Comment (tpluscode) [inline apis/core/lib/domain/dataset/update.ts:46]: You don't need to modify this file. The dataset resource is in a different graph from the metadata

**Review by tpluscode — CHANGES_REQUESTED** [review]
You can now rebase master. The `createIdentifier` got merged

File: apis/core/lib/domain/dataset/update.ts (hunk 9 above)
Comment (lucafurrer) [reply apis/core/lib/domain/dataset/update.ts:46]: But the triple `dataset cc.dimensionMetadata metadatacollection` has been deleted by the deleteOut by an update in the end to end tests.

Comment (tpluscode) [reply apis/core/lib/domain/dataset/update.ts:46]: Aha, my mistake. I mistook these for the `schema:hasPart` triples. Sorry

**Comment by lucafurrer** [comment]
I think I addressed all the points

File: packages/model/DimensionMetadata.ts [hunk 10]
```diff
@@ -42,3 +42,9 @@ DimensionMetadataCollectionMixin.appliesTo = cc.DimensionMetadataCollection
 export const createCollection = initializer<DimensionMetadataCollection>(DimensionMetadataCollectionMixin, {
   types: [cc.DimensionMetadataCollection],
 })
+
+DimensionMetadataCollectionMixin.appliesTo = cc.DimensionMetadataCollection
```
Comment (tpluscode) [inline packages/model/DimensionMetadata.ts:46]: This got duplicated in merge

File: apis/core/lib/domain/dimension-metadata/DimensionMetadata.ts [hunk 11]
```diff
@@ -0,0 +1,46 @@
+import { Constructor } from '@tpluscode/rdfine'
```
Comment (tpluscode) [inline apis/core/lib/domain/dimension-metadata/DimensionMetadata.ts:1]: Uh, oh., I created the same model in `apis/core/lib/domain/dimension` directory. Sorry 😊 

Please merge `dimension` and `dimension-metadata`. Flip a coin to decide on the name :)

File: apis/core/lib/domain/dimension-metadata/DimensionMetadata.ts [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+  interface DimensionMetadataCollection extends ApiDimensionMetadataCollection {
+  }
+}
+
+export default function Mixin<Base extends Constructor<DimensionMetadataCollection>>(Resource: Base) {
+  return class Impl extends Resource implements ApiDimensionMetadataCollection {
+    addDimensionMetadata(params: CreateColumnMetadata): DimensionMetadata.DimensionMetadata {
+      const identifier = params.csvMapping.createIdentifier(params.columnMapping.targetProperty)
+
+      if (this.hasPart && this.hasPart.some((dimMeta) => dimMeta.about === identifier)) {
```
Comment (tpluscode) [inline apis/core/lib/domain/dimension-metadata/DimensionMetadata.ts:31]: As discussed on zulip, checking `this.hasPart` for truthiness will not be necessary

File: apis/core/lib/domain/dimension-metadata/DimensionMetadata.ts [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+        throw new Error(`Dimension Metadata with identifier ${identifier.value} already exists`)
+      }
+
+      const dimensionMetadata = DimensionMetadata.create(this.pointer.node(id.dimensionMetadata(this, params.columnMapping)), {
```
Comment (tpluscode) [inline apis/core/lib/domain/dimension-metadata/DimensionMetadata.ts:35]: I've drifted a little away from really needing the `identifiers` module. In this case for example, `id.dimensionMetadata` will likely only ever be used here. 

I propose to move into this module instead

File: apis/core/lib/domain/identifiers.ts [hunk 14]
```diff
@@ -55,3 +56,12 @@ export function user(user: string): NamedNode {
...
 }
+
+export function dimensionMetadata(dimensionMetadataCollection: DimensionMetadataCollection, mapping: ColumnMapping): NamedNode {
+  if (mapping.targetProperty.termType === 'Literal') {
+    return $rdf.namedNode(`${dimensionMetadataCollection.id.value}/${mapping.targetProperty.value}`)
+  } else {
+    const prefixedProperty = shrink(mapping.targetProperty.value) || mapping.targetProperty.value
+    return $rdf.namedNode(`${dimensionMetadataCollection.id.value}/${prefixedProperty}`)
+  }
+}
```
Comment (tpluscode) [inline apis/core/lib/domain/identifiers.ts:67]: 1. May be slightly more readable if there was only one `return`?
2. Please `urlencode` the URI if did not shrink

**Review by tpluscode — CHANGES_REQUESTED** [review]

**Review by tpluscode — APPROVED** [review]

**Pull request merged**